### ISSUE TITLE
Readability of DSL2 pipelines / documentation of pipeline structures. 

### ISSUE THREAD

**Issue by grst** [nf-core/tools#1163]
Since we touched the topic today in #1159, here's my story: 

Say I have no previous experience with nf-core pipelines and want to find out what parameters `STAR` runs with. Here's my odysee: 

1. I start with reading [main.nf](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/main.nf). It [includes the `RNASEQ` workflow](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/main.nf#L43). Ok...
2. [workflows/rnaseq.nf](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/workflows/rnaseq.nf) handles and amazing amount of configuration options. Somewhere far down, I find an include statement for `ALIGN_STAR`. Let's see...
3. The [align_star subworkflow](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/subworkflows/nf-core/align_star.nf) doesn't seem to do a lot, but I find out that it loads the module `STAR_ALIGN`
4. Finally, [I found the command](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/modules/nf-core/software/star/align/main.nf#L48) that runs `STAR`. Unfortunately, the parameters are just passed from a variable called `$options.args`. Which comes from a function `initOptions(params.options)`. `params.options` is initalized as an empty map. 
5. So, let's check what [`initOptions`](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/modules/nf-core/software/star/align/functions.nf#L15) does. Apparently not a lot, it just sets empty defaults for a few keys. 
6. So where do the params actually come from? By asking nicely on slack (or reading the [nextflow documentation](https://www.nextflow.io/docs/latest/dsl2.html#module-parameters)) I learn that `params` gets overwritten upon module import.  
7. Therefore, I'm back at [`align_star.nf`](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/subworkflows/nf-core/align_star.nf). `STAR_ALIGN` is included with `params.align_options`. Apparently, `params` also gets overwritten in subworkflows...
8. Therefore, I again check [`rnaseq.nf`](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/workflows/rnaseq.nf). By scrolling horizontally in a 309-character line (WTF?) I learn that STAR is called with `star_align_options`. `star_align_options` is manipulated a little bit further up in the same file, and I find out it is actually initalized from `modules['star_align']`. Scrolling to [the top (well, line 74/703)](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/workflows/rnaseq.nf#L74), I finally learn that `modules` is copied from `params.modules.clone()`. Since `params` is not defined in this file, I check the config files. 
9. I already know that nextflow checks the [`nextflow.config`](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/nextflow.config). It doesn't include anything about `modules`, but it [does include](https://github.com/nf-core/rnaseq/blob/b3ff92bc54363faf17d820689a8e9074ffd99045/nextflow.config#L117) `conf/modules.config`. 
10. This, finally, seems the place I have been looking for :tada:. STAR get's called with `--quantMode TranscriptomeSAM --twopassMode Basic --outSAMtype BAM Unsorted --readFilesCommand zcat --runRNGseed 0 --outFilterMultimapNmax 20 --alignSJDBoverhangMin 1 --outSAMattributes NH HI AS NM MD --quantTranscriptomeBan Singleend`. 

Improving documentation and making an overview figure as suggested by @aunderwo, is definitely a good start. 
Limiting line-length to 90-120 characters probably a good idea. But if there are other suggestions how to optimize the pipeline structure for readability, I'm certainly happy to hear them. 

Ping @KevinMenden, @drpatelh, @ewels 
CC @riederd

**Comment by drpatelh** [comment]
As commented on Slack:

I agree it isn't straightforward to get to grips with some of this stuff but I also feel it's one of those things that you have to spend a little time staring at to understand it all properly. I think I have written my fair share of DSL2 documentation now! So if anyone else fancies chipping in that would be fab :sweat_smile:

## Pipeline docs

I also wrote some [docs](https://nf-co.re/rnaseq/3.2/usage#custom-configuration) for nf-core/rnaseq and nf-core/viralrecon that will eventually be added to pipelines via the template too. These explain how users can tweak DSL2 configuration options for resources as well as how to overwrite module-specific options via a config file.
## Bytesize talk no. 5

I tried to explain this all conceptually in the gargantuan #bytesize talk I gave which is available on [Youtube](https://www.youtube.com/watch?v=ggGGhTMgyHI)

## Extracts from Bytesize talk no. 5

This is the slide I came up with in the talk above to explain passing around module options. The figure isn't all encompassing but was intentionally written to be simplistic enough for beginners to understand:

![image](https://user-images.githubusercontent.com/23529759/124640271-176d7c80-de85-11eb-88bf-39722e6f7b07.png)

> By scrolling horizontally in a 309-character line (WTF?) 

I am not :100: with it either to be honest but then the code wrapping gets really ugly when you have loads of options so I just went with "Meh". The downside with code wrapping specifically for module options (which can be quite long) is that your end up scrolling through endless lines to get to where you want to be even though you may not be editing these sections of the code that often 🤷🏽

**Comment by grst** [comment]
I had another look at the RNA-seq pipeline and a comment and a question: 
 * Actually the workflows themselves are super-readable with DSL2. It is really just the configuration(-flow) that I find hard to follow. 
 * What was the motivation of having a separate `workflows` directory instead of just putting the main workflow in `main.nf`? This feels like an unnecessary layer of abstraction. 

Ongoing stuff: 

* An overview figure (or rather two - one for configuration and one for directory structure) is now being prepared by @aunderwo  as [discussed on slack](https://nfcore.slack.com/archives/CJRH30T6V/p1625581147127800). 
 
Code-style: 

> By scrolling horizontally in a 309-character line (WTF?)

My suggested formatting would be the following. It easily fits and a line, and results in meaningful diffs when changing only a single option. 
It is true, though, that it significantly increases the line count, but then again, these kind of blocks can 
be easily collapsed in vscode: 

![Peek 2021-07-13 15-11](https://user-images.githubusercontent.com/7051479/125457598-6e7ea0be-7ee9-4dbe-98cb-7ced21adb9c3.gif)

**Comment by mahesh-panchal** [comment]
I just want to leave here that I made a start on writing some info on how to use nf-core modules here. 
https://github.com/mahesh-panchal/nf-core_modules_demo

My plan is to eventually write something to include on the website in the (hopefully near) future. 
I just saw that the modules format has changed so I need to update it to reflect that.

**Comment by grst** [comment]
I consider this solved with DSL2 2.0 :) 

But can we maybe still get a one of @jfy133's awesome figures showing how the new system works?

**Comment by jfy133** [comment]
If you make a rough sketch (e.g. on paper), I'm happy to make it all spiffy!

**Comment by drpatelh** [comment]
Let's close this then guys. Thanks!
